lib/io/can.cpp

```diff
@@ -9,111 +9,113 @@
   return io::CanResult::kSuccess;
 }
 
 std::optional<CanFrame> Can::receive()
 {
   CanFrame received_message;
   if (ioctl(socket_, FIONREAD) < sizeof(CanFrame)) {
-    logger_.log(core::LogLevel::kDebug, "No can message in rx queue");
+    log_.log(core::LogLevel::kDebug, "No can data in rx queue");
     return std::nullopt;
   }
   const int num_bytes_read = read(socket_, &received_message, sizeof(CanFrame));
   if (num_bytes_read < sizeof(CanFrame)) {
-    logger_.log(core::LogLevel::kFatal, "Failed to receive CAN message");
+    log_.log(core::LogLevel::kFatal, "Failed to receive CAN data");
     return std::nullopt;
   }
-  // TODOLater make logger more elegant
-  logger_.log(core::LogLevel::kDebug,
-              "CAN message received, ID:%i, DATA: %i %i %i %i %i %i %i %i",
-              static_cast<int>(received_message.can_id),
-              static_cast<int>(received_message.data[0]),
-              static_cast<int>(received_message.data[1]),
-              static_cast<int>(received_message.data[2]),
-              static_cast<int>(received_message.data[3]),
-              static_cast<int>(received_message.data[4]),
-              static_cast<int>(received_message.data[5]),
-              static_cast<int>(received_message.data[6]),
-              static_cast<int>(received_message.data[7]));
+  // TODOLater make log_ more elegant
+  log_.log(core::LogLevel::kDebug,
+           "CAN data received, ID:%i, DATA: %i %i %i %i %i %i %i %i",
+           static_cast<int>(received_message.can_id),
+           static_cast<int>(received_message.data[0]),
+           static_cast<int>(received_message.data[1]),
+           static_cast<int>(received_message.data[2]),
+           static_cast<int>(received_message.data[3]),
+           static_cast<int>(received_message.data[4]),
+           static_cast<int>(received_message.data[5]),
+           static_cast<int>(received_message.data[6]),
+           static_cast<int>(received_message.data[7]));
 
   return received_message;
 }
 
-void Can::listen()
+CanResult Can::listen()
 {
-  std::optional<CanFrame> data;
-  if (ioctl(socket_, FIONREAD) >= sizeof(CanFrame)) { data = receive(); }
-  if (data == std::nullopt) { return; }
-  const CanFrame message = data.value();
+  if (ioctl(socket_, FIONREAD) < sizeof(CanFrame)) { return io::CanResult::kFailure; }
+  const auto data = receive();
+  if (!data) { return io::CanResult::kFailure; }
+  CanFrame message = data.value();
   if (!processors_.contains(message.can_id)) {
-    logger_.log(core::LogLevel::kInfo, "No CanProccessor associated with id %i", message.can_id);
+    log_.log(core::LogLevel::kInfo, "No CanProccessor associated with id %i", message.can_id);
+    return io::CanResult::kFailure;
   }
-  for (uint8_t i = 0; i < processors_[message.can_id].size(); i++) {
-    processors_[message.can_id][i]->processMessage(message);
+  std::vector<std::shared_ptr<ICanProcessor>> &subscribed_processors = processors_[message.can_id];
```

Comment:
again, what if `message.can_id` isn't in the map? use `std::unordered_map::find`!

Reply:
isn't this already caught by if (!processors_.contains(message.can_id)) on line 104?

Reply:
oh, you're right. my bad. here's the idiomatic way to do it (fewer lookups; also `std::map::contains` didn't exist pre C++20 so you always had to do this)

```cpp
auto it = processors_.find(message.can_id);
if (it == processors.end()) {
   // handle error
}
for (auto &processor : it->second) { 
  // do stuff
}
```

I don't think we like the variable name `it` here so call it something like ` canIdAndProcessor`.

Reply:
It also might be possible to sprinkle in some `const`s here; not sure whether you can put it on the iterator because the semantics of what is and isn't immutable due to const are a bit fuzzy. try it out and see what works.

Reply:
i believe the iterator is alright, it builds and i can't see anything obvious online on why it wouldn't work, but i'm not 100% confident